src/main/java/org/cru/redegg/recording/impl/DefaultStuckThreadMonitor.java

```diff
@@ -0,0 +1,260 @@
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@ApplicationScoped
+public class DefaultStuckThreadMonitor implements StuckThreadMonitor
+{
+
+    private static final Logger LOG = LoggerFactory.getLogger(DefaultStuckThreadMonitor.class);
+
+    /*
+     * Use a weak-keys map so that, on the off chance that a thread doesn't call
+     * finishMonitoringRequest() before it finishes a request
+     * (some kind of Error might be thrown, for example),
+     * the map entry will eventually get garbage collected.
+     */
+    private final Map<WebContext, Request> currentlyActiveRequests = new MapMaker()
+        .initialCapacity(50)
+        .weakKeys()
+        .makeMap();
+
+    private final ScheduledExecutorService executorService =
+        new ScheduledThreadPoolExecutor(1);
+
+
+    private ReadablePeriod threshold;
+    private long period;
+    private TimeUnit periodTimeUnit;
+
+    private InetAddress localHost;
+    {
+        try
+        {
+            localHost = InetAddress.getLocalHost();
+        }
+        catch (UnknownHostException e)
+        {
+            LOG.warn("can't get localhost", e);
+        }
+    }
+
+
+    private final Clock clock;
+    private final ErrorQueue errorQueue;
+
+    @Inject
+    public DefaultStuckThreadMonitor(
+        Clock clock,
+        ErrorQueue errorQueue,
+        @Selected StuckThreadMonitorConfig config)
+    {
+        this.clock = clock;
+        this.errorQueue = errorQueue;
+        this.threshold = config.getThreshold();
+        this.period = config.getPeriod();
+        this.periodTimeUnit = config.getPeriodTimeUnit();
+    }
+
+    @ProxyConstructor
+    public DefaultStuckThreadMonitor()
+    {
+        this.clock = null;
+        this.errorQueue = null;
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        executorService.scheduleAtFixedRate(
+            new Scanner(),
+            period,
+            period,
+            periodTimeUnit);
+    }
+
+    @Override
+    public void startMonitoringRequest(WebContext webContext)
+    {
+        Request request = new Request(
+            webContext.getStart().plus(threshold),
+            webContext.clone(),
+            Thread.currentThread());
+
+        currentlyActiveRequests.put(webContext, request);
+    }
+
+    @Override
+    public void finishMonitoringRequest(WebContext webContext)
+    {
+        currentlyActiveRequests.remove(webContext);
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        executorService.shutdown();
+        try
+        {
+            executorService.awaitTermination(5, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e)
+        {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("interrupted while awaiting termination", e);
```

Comment:
Interesting that you are throwing a `RuntimeException` here, I usually just set the interrupt flag and get out of my own code as quickly as possible.

Throwing a RuntimeException means that the caller logic which we don't control won't have a chance to perform any additional cleanup that isn't blocking. and the caller won't know the actual reason we are bailing so quickly.

Reply:
Yeah, I don't know the best way to handle this case. It's not a likely case, so I didn't spend much time thinking about it.

I think you're right, though. And actually I did what you suggest in another place. I'll extract a helper from it.

Reply:
Ok, fixed/squashed.
Also, I added a shutdownNow() call.